aws/credentials/ibmiam/trusted_profile_credentials_provider.go

```diff
@@ -0,0 +1,168 @@
+
+	logLevel := aws.LogLevel(aws.LogOff)
+	if config != nil && config.LogLevel != nil && config.Logger != nil {
+		logLevel = config.LogLevel
+		provider.logger = config.Logger
+	}
+	provider.logLevel = logLevel
+
+	if crTokenFilePath == "" {
+		provider.ErrorStatus = awserr.New("crTokenFilePathNotFound", "CR token file path not found", nil)
+		if provider.logLevel.Matches(aws.LogDebug) {
+			provider.logger.Log(debugLog, "<IBM IAM PROVIDER BUILD>", provider.ErrorStatus)
+		}
+
+		return provider
+	}
+
+	if trustedProfileName == "" && trustedProfileID == "" {
+		provider.ErrorStatus = awserr.New("trustedProfileNotFound", "Trusted profile name or id not found", nil)
+		if provider.logLevel.Matches(aws.LogDebug) {
+			provider.logger.Log(debugLog, "<IBM IAM PROVIDER BUILD>", provider.ErrorStatus)
+		}
+
+		return provider
+	}
+
+	if authEndPoint == "" {
+		authEndPoint = defaultAuthEndPoint
+		if provider.logLevel.Matches(aws.LogDebug) {
+			provider.logger.Log(debugLog, "<IBM IAM PROVIDER BUILD>", "using default auth endpoint", authEndPoint)
+		}
+	}
+
+	authenticator, err := core.NewContainerAuthenticatorBuilder().
+		SetIAMProfileName(trustedProfileName).
+		SetIAMProfileID(trustedProfileID).
+		SetCRTokenFilename(crTokenFilePath).
+		SetURL(authEndPoint).
+		Build()
+	if err != nil {
+		provider.ErrorStatus = awserr.New("errCreatingAuthenticatorClient", "cannot setup new Authenticator client", err)
+		if provider.logLevel.Matches(aws.LogDebug) {
+			provider.logger.Log(debugLog, "<IBM IAM PROVIDER BUILD>", provider.ErrorStatus)
+		}
+
+		return provider
+	}
+
+	provider.authenticator = authenticator
+
+	return provider
+}
+
+// IsValid ...
+// Returns:
+//
+//	TrustedProfileProvider validation - boolean
+func (p *TrustedProfileProvider) IsValid() bool {
+	return nil == p.ErrorStatus
+}
+
+// Retrieve ...
+// Returns:
+//
+//	Credential values
+//	Error
+func (p *TrustedProfileProvider) Retrieve() (credentials.Value, error) {
+	if p.ErrorStatus != nil {
+		if p.logLevel.Matches(aws.LogDebug) {
+			p.logger.Log(debugLog, ibmiamProviderLog, p.providerName, p.ErrorStatus)
+		}
+		return credentials.Value{ProviderName: p.providerName}, p.ErrorStatus
+	}
+
+	tokenValue, err := p.authenticator.GetToken()
+	if err != nil {
+		var returnErr error
+		if p.logLevel.Matches(aws.LogDebug) {
+			p.logger.Log(debugLog, ibmiamProviderLog, p.providerName, "ERROR ON GET token", err)
+			returnErr = awserr.New("TokenRetrieveError", "error retrieving the token", err)
+		} else {
+			returnErr = awserr.New("TokenRetrieveError", "error retrieving the token", nil)
+		}
+		return credentials.Value{}, returnErr
+	}
+
+	return credentials.Value{
+		Token: token.Token{
+			AccessToken: tokenValue,
+			TokenType:   tokenType,
+		},
+		ProviderName: p.providerName,
+		ProviderType: p.providerType,
+	}, nil
+}
+
+// IsExpired ...
+//
+//	TrustedProfileProvider expired or not - boolean
+func (p *TrustedProfileProvider) IsExpired() bool {
+	return true
```

Comment:
why it is always true but not false?

Reply:
The [GetToken](https://github.com/IBM/ibm-cos-sdk-go/pull/12/files#diff-1484b4e912fa22bc37444fa03e4c20af0838c14881dbce4eb0dae7e66199111bR129) function is actually checking the expiration before it's trying to get the token from server. Also the token in authenticator is not accessible in `IsExpired` function since it is not an exported variable and we won't be able to check the expiration from our side. I think it should be okay to return true since that logic is handled inside the `GetToken` function. Also updated the doc comment.